src/test/java/com/library/controller/BookControllerTest.java

```diff
@@ -0,0 +1,207 @@
+                        .param("page", "1")
+                        .param("sizeOnPage", "5"))
+                .andDo(print())
+                .andExpect(status().isOk())
+                .andReturn()
+                .getResponse();
+        JSONArray jsonArray = new JSONArray(response.getContentAsString());
+        JSONObject jsonObject = jsonArray.getJSONObject(0);
+
+        assertNotNull(response.getContentAsString());
+        assertEquals(1, jsonObject.get("id"));
+        assertEquals("War and peace", jsonObject.get("title"));
+        assertEquals(3, jsonArray.length());
+        assertEquals("application/json", response.getContentType());
+        verify(bookRepository).findAll(paging);
+    }
+
+    @SneakyThrows
+    @Test
+    void getBookById() {
+        when(bookRepository.findById(1L)).thenReturn(Optional.of(testStorage.getBook()));
+
+        MockHttpServletResponse response = mockMvc.perform(get("/books/{id}", "1"))
+                .andDo(print())
+                .andExpect(status().isOk())
+                .andReturn()
+                .getResponse();
+        JSONObject jsonObject = new JSONObject(response.getContentAsString());
+
+        assertEquals("update", jsonObject.get("title"));
+        assertEquals(1, jsonObject.get("id"));
+        verify(bookRepository).findById(1L);
+    }
+
+    @SneakyThrows
+    @Test
+    void getBookByIdFailed() {
+        when(bookRepository.findById(4L)).thenThrow(EntityNotFoundException.class);
+
+        MockHttpServletResponse response = mockMvc.perform(get("/books/{id}", "4"))
+                .andDo(print())
+                .andExpect(status().isNotFound())
+                .andReturn()
+                .getResponse();
+        JSONObject jsonObject = new JSONObject(response.getContentAsString());
+
+        assertEquals("null", jsonObject.get("message").toString());
+    }
+
+    @SneakyThrows
+    @Test
+    void saveBook() {
+        Book newBook = testStorage.getNewBook();
+        BookDto bookDto = bookMapper.mapToBookDto(newBook);
+
+        when(bookRepository.save(argThat(new IsSameLikeBook(newBook)))).thenReturn(newBook);
+
+        MockHttpServletResponse response = mockMvc.perform(post("/books")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(bookDto)))
+                .andExpect(status().isOk())
+                .andReturn()
+                .getResponse();
+
+        JSONObject jsonObject = new JSONObject(response.getContentAsString());
+
+        assertNotNull(response);
+        assertEquals(testStorage.getBook().getTitle(), jsonObject.get("title"));
+        verify(bookRepository).save(argThat(new IsSameLikeBook(newBook)));
+    }
+
+    @SneakyThrows
+    @Test
+    void saveBookFailed() {
+        Book newBook = testStorage.getNewBook();
+        newBook.setAuthors(null);
+        BookDto bookDto = bookMapper.mapToBookDto(newBook);
+
+
```

Comment:
1 - Extra line

2 - Where is mock for repository?

Reply:
1- ok
2 - Why is it needed here? App fails in service, repository is not used